zetaclient/chains/bitcoin/tx_script.go

```diff
@@ -169,27 +166,49 @@ func DecodeScriptP2PKH(scriptHex string, net *chaincfg.Params) (string, error) {
 
 // DecodeOpReturnMemo decodes memo from OP_RETURN script
 // returns (memo, found, error)
-func DecodeOpReturnMemo(scriptHex string, txid string) ([]byte, bool, error) {
-	if len(scriptHex) >= 4 && scriptHex[:2] == "6a" { // OP_RETURN
-		memoSize, err := strconv.ParseInt(scriptHex[2:4], 16, 32)
-		if err != nil {
-			return nil, false, errors.Wrapf(err, "error decoding memo size: %s", scriptHex)
+func DecodeOpReturnMemo(scriptHex string) ([]byte, bool, error) {
+	// decode hex script
+	scriptBytes, err := hex.DecodeString(scriptHex)
+	if err != nil {
+		return nil, false, errors.Wrapf(err, "error decoding script hex: %s", scriptHex)
+	}
+
+	// skip non-OP_RETURN script
+	// OP_RETURN script has to be at least 2 bytes: [OP_RETURN + dataLen]
+	if len(scriptBytes) < 2 || scriptBytes[0] != txscript.OP_RETURN {
+		return nil, false, nil
+	}
+
+	// extract appended data in the OP_RETURN script
+	var memoBytes []byte
+	var memoSize = scriptBytes[1]
+	switch {
+	case memoSize < txscript.OP_PUSHDATA1:
+		// memo size has to match the actual data
+		if int(memoSize) != (len(scriptBytes) - 2) {
+			return nil, false, fmt.Errorf("memo size mismatch: %d != %d", memoSize, (len(scriptBytes) - 2))
 		}
-		if int(memoSize) != (len(scriptHex)-4)/2 {
-			return nil, false, fmt.Errorf("memo size mismatch: %d != %d", memoSize, (len(scriptHex)-4)/2)
+		memoBytes = scriptBytes[2:]
+	case memoSize == txscript.OP_PUSHDATA1:
+		// when data size >= OP_PUSHDATA1 (76), Bitcoin uses 2 bytes to represent the length: [OP_PUSHDATA1 + dataLen]
+		// see: https://github.com/btcsuite/btcd/blob/master/txscript/scriptbuilder.go#L183
+		if len(scriptBytes) < 3 {
+			return nil, false, fmt.Errorf("script too short: %s", scriptHex)
 		}
+		memoSize = scriptBytes[2]
```

Comment:
nit: for my own education, on this case shouldn't memoSize be scriptBytes[1:2], that is the bytes 1 and 2, since btc will use 2 bytes to represent the length.

Reply:
This part is a trick. They fill in the 2 bytes in format `[OP_PUSHDATA1][dataLen]`, in which the 1st byte is served as a flag. When flag is set to `OP_PUSHDATA1`, we need to take the 2nd byte only. Their code is [here](https://github.com/btcsuite/btcd/blob/master/txscript/scriptbuilder.go#L183)